aidial_adapter_bedrock/llm/model/llama/v3.py

```diff
@@ -1,72 +1,38 @@
-"""
-Turning a chat into a prompt for the Llama3 model.
-
-See as a reference:
-https://github.com/meta-llama/llama3/blob/main/llama/tokenizer.py
-"""
-
-from typing import List, Literal, Optional, Tuple, assert_never
-
-from aidial_adapter_bedrock.llm.chat_emulator import ChatEmulator
-from aidial_adapter_bedrock.llm.message import (
-    AIRegularMessage,
-    BaseMessage,
-    HumanRegularMessage,
-    SystemMessage,
+import json
+from typing import Any, List, Tuple
+
+from aidial_adapter_bedrock.dial_api.request import ModelParameters
+from aidial_adapter_bedrock.llm.converse.adapter import ConverseAdapter
+from aidial_adapter_bedrock.llm.converse.types import (
+    ConverseDeployment,
+    ConverseMessage,
+    ConverseParams,
 )
-from aidial_adapter_bedrock.llm.model.llama.conf import LlamaConf
-
-
-def get_role(message: BaseMessage) -> Literal["system", "user", "assistant"]:
-    match message:
-        case SystemMessage():
-            return "system"
-        case HumanRegularMessage():
-            return "user"
-        case AIRegularMessage():
-            return "assistant"
-        case _:
-            assert_never(message)
-
+from aidial_adapter_bedrock.llm.tokenize import default_tokenize_string
 
-def encode_header(message: BaseMessage) -> str:
-    ret = ""
-    ret += "<|start_header_id|>"
-    ret += get_role(message)
-    ret += "<|end_header_id|>"
-    ret += "\n\n"
-    return ret
 
+class ConverseStreamingEmulateAdapter(ConverseAdapter):
+    """
+    Llama 3.1 supports tools, but only in non-streaming mode.
+    So we need to run request in non-streaming mode, and then emulate streaming.
+    """
 
-def encode_message(message: BaseMessage) -> str:
-    ret = encode_header(message)
-    ret += message.text_content.strip()
-    ret += "<|eot_id|>"
-    return ret
+    def is_stream(self, params: ModelParameters) -> bool:
+        if self.get_tool_config(params):
+            return False
+        return params.stream
 
 
-def encode_dialog_prompt(messages: List[BaseMessage]) -> str:
-    ret = ""
-    ret += "<|begin_of_text|>"
-    for message in messages:
-        ret += encode_message(message)
-    ret += encode_header(AIRegularMessage(content=""))
-    return ret
+def input_tokenizer_factory(
+    deployment: ConverseDeployment, params: ConverseParams
+):
+    tool_tokens = default_tokenize_string(json.dumps(params.toolConfig))
```

Comment:
Should you also add tokens for `params.system`?

Reply:
Fixed